sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dofn/ReadChangeStreamPartitionDoFn.java

```diff
@@ -146,6 +156,17 @@ public TimestampRange initialRestriction(@Element PartitionMetadata partition) {
     return TimestampRange.of(startTimestamp, endTimestamp);
   }
 
+  @GetSize
+  public double getSize(@Element PartitionMetadata partition, @Restriction TimestampRange range)
+      throws Exception {
+    final BigDecimal timeGapInSeconds =
+        BigDecimal.valueOf(newTracker(partition, range).getProgress().getWorkRemaining());
+    final BigDecimal throughput = BigDecimal.valueOf(this.throughputEstimator.get());
+    LOG.debug(
+        "Reported getSize() - remaining work: " + timeGapInSeconds + " throughput:" + throughput);
+    return timeGapInSeconds.multiply(throughput).doubleValue();
```

Comment:
Any chance of an overflow here? Could you add property based tests for this?

Reply:
To make it easy for testing, I have to change to the following code: 

```
  @GetSize
  public double getSize(@Element PartitionMetadata partition, @Restriction TimestampRange range)
      throws Exception {
    return getBacklogSize(newTracker(partition, range).getProgress().getWorkRemaining(), this.throughputEstimator.get());
  }

  @VisibleForTesting
  private double getBacklogSize(double workRemaining, double localThroughput) {
    final BigDecimal timeGapInSeconds = BigDecimal.valueOf(workRemaining);
    final BigDecimal throughput = BigDecimal.valueOf(localThroughput);
    LOG.debug(
        "Reported getSize() - remaining work: " + timeGapInSeconds + " throughput:" + throughput);
    return timeGapInSeconds.multiply(throughput).doubleValue();
  }
```

It has a chance of an overflow, e.g., `throughput` is the max of double although `timeGapInSeconds` should have a lower range (getSeconds() returns long) but it will have an overflow. I wonder the property based tests would have a meaning here because we are trying to test multiplying a double with a long. 

To avoid an overflow, we probably can change it to (avoid the overflow):  

```
return timeGapInSeconds.multiply(throughput).min(BigDecimal.valueOf(Double.MAX_VALUE)).doubleValue();
```